### PR TITLE
[NNCF] Update test nncf_graph 

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
* Add feature to generate reference graph by set the global variable `NNCF_TEST_REGEN_DOT`
* Add reference graphs for anomaly models
* Task: 85980, 85990, 85991. 85992

### CODE REVIEW

**Comment by AlexanderDokuchaev** [comment]
run ote-test

File: external/anomaly/tests/anomaly_common.py [hunk 1]
```diff
@@ -107,3 +107,19 @@ def get_anomaly_domain_test_action_classes(
         OTETestNNCFExportEvaluationAction,
         OTETestNNCFGraphAction,
     ]
+
+
+def get_dummy_compressed_model(task):
```
Comment (samet-akcay) [inline external/anomaly/tests/anomaly_common.py]: Would it be possible to annotate the args with type?

File: ote_sdk/ote_sdk/test_suite/training_tests_actions.py [hunk 2]
```diff
@@ -562,15 +562,23 @@ def __call__(self, data_collector: DataCollector, results_prev_stages: OrderedDi
 
 
 # TODO: think about move to special file
-def check_nncf_model_graph(model, path_to_dot):
+def check_nncf_model_graph(model, path_to_dot, force_regen_dot=False):
```
Comment (samet-akcay) [inline ote_sdk/ote_sdk/test_suite/training_tests_actions.py]: Here as well. Would be good to know the arg and return types, which improves the readability.

**Review by samet-akcay — COMMENTED** [review]
Thanks @AlexanderDokuchaev. I have just few comments regarding the PR.

I've another concern regarding the integration with the otx branch, though. We need to think about how/where we merge these changes in the `feature/otx` [branch](https://github.com/openvinotoolkit/training_extensions/tree/feature/otx). 

@ashwinvaidya17, maybe we could discuss this?

File: external/anomaly/tests/anomaly_common.py (hunk 1 above)
Comment (AlexanderDokuchaev) [reply external/anomaly/tests/anomaly_common.py]: done

File: ote_sdk/ote_sdk/test_suite/training_tests_actions.py (hunk 2 above)
Comment (AlexanderDokuchaev) [reply ote_sdk/ote_sdk/test_suite/training_tests_actions.py]: done

**Comment by AlexanderDokuchaev** [comment]
Changed target branch

File: external/anomaly/tests/reference/ote_anomaly_classification_padim/nncf/nncf_quantization.dot [hunk 3]
```diff
@@ -0,0 +1,240 @@
+strict digraph  {
```
Comment (ashwinvaidya17) [inline external/anomaly/tests/reference/ote_anomaly_classification_padim/nncf/nncf_quantization.dot:1]: Why are these dot files added? Can't we regenerate these from the tests?

File: external/anomaly/tasks/inference.py [hunk 4]
```diff
@@ -291,8 +291,11 @@ def save_model(self, output_model: ModelEntity) -> None:
     def _set_metadata(self, output_model: ModelEntity):
         output_model.set_data("image_threshold", self.model.image_threshold.value.cpu().numpy().tobytes())
         output_model.set_data("pixel_threshold", self.model.pixel_threshold.value.cpu().numpy().tobytes())
-        output_model.set_data("min", self.model.min_max.min.cpu().numpy().tobytes())
-        output_model.set_data("max", self.model.min_max.max.cpu().numpy().tobytes())
+        if hasattr(self.model, "min_max"):
+            output_model.set_data("min", self.model.min_max.min.cpu().numpy().tobytes())
+            output_model.set_data("max", self.model.min_max.max.cpu().numpy().tobytes())
+        else:
+            logger.warning("Model saved without min_max")
```
Comment (ashwinvaidya17) [inline external/anomaly/tasks/inference.py]: Maybe we should mention that model wasn't initialized properly.

File: ote_sdk/ote_sdk/test_suite/training_tests_actions.py [hunk 5]
```diff
@@ -562,15 +565,25 @@ def __call__(self, data_collector: DataCollector, results_prev_stages: OrderedDi
 
 
 # TODO: think about move to special file
-def check_nncf_model_graph(model, path_to_dot):
+def check_nncf_model_graph(
+    model: "NNCFNetwork", path_to_dot: str, force_regen_dot: bool = False
```
Comment (ashwinvaidya17) [inline ote_sdk/ote_sdk/test_suite/training_tests_actions.py:569]: Why is `force_regen_dot` set to `False`? Personally, I am in the favour of avoiding committing `dot` files to the repo. The tests might fail if the model architecture changes or if a new model is added with missing dot files.

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks. I have a few comments.

File: external/anomaly/tests/reference/ote_anomaly_classification_padim/nncf/nncf_quantization.dot (hunk 3 above)
Comment (AlexanderDokuchaev) [reply external/anomaly/tests/reference/ote_anomaly_classification_padim/nncf/nncf_quantization.dot:1]: No. The idea of test to compare graph of model with reference that generated when it works correctly. If graph changes without update nncf configs, it will be unwanted behavior. It's like reference evaluation metric, you want to compare current state with reference.

Comment (ashwinvaidya17) [reply external/anomaly/tests/reference/ote_anomaly_classification_padim/nncf/nncf_quantization.dot:1]: Okay that makes sense. But it is something for us to keep in mind when we make any changes to the model.

File: ote_sdk/ote_sdk/test_suite/training_tests_actions.py (hunk 5 above)
Comment (AlexanderDokuchaev) [reply ote_sdk/ote_sdk/test_suite/training_tests_actions.py:569]: > The tests might fail if the model architecture changes or if a new model is added with missing dot files.

Yes, it was the main reason to add this test.


The test added to indicate that model was changed. It's matter because optimization configs was added for specific model and if model will be changes optimization can work incorrectly.


`force_regen_dot` added to simplify method to update reference graph, it does not change the test. It's works before but to generate the reference graph we need to add some code. Now it can be done by set environment variable.

File: external/anomaly/tasks/inference.py (hunk 4 above)
Comment (AlexanderDokuchaev) [reply external/anomaly/tasks/inference.py]: I think no, because a model was initialized properly. It happens if save a model without running training pipeline. 
Can you suggest more correctly warning message?

Comment (ashwinvaidya17) [reply external/anomaly/tasks/inference.py]: Min-Max is assigned when `setup` is called. It is initialized when the training pipeline is run. If we just create a model and do not call the train loop, it will in consequence not initialize the required classes used for normalization. The output produced in this case will not be good.

Comment (ashwinvaidya17) [reply external/anomaly/tasks/inference.py]: Maybe we can say something like `the model was not trained before saving. This will lead to incorrect normalization of the heatmaps.`

Comment (AlexanderDokuchaev) [reply external/anomaly/tasks/inference.py]: Updated

Comment (ashwinvaidya17) [reply external/anomaly/tasks/inference.py]: Thanks

**Review by ashwinvaidya17 — COMMENTED** [review]
Looks good to me. I'll wait for @samet-akcay 's suggestions before approving this PR

**Comment by samet-akcay** [comment]
Just to give @AlexanderDokuchaev a little background, we have been working on `OTE v2`, of which the first phase will soon be merged.

The tricky part in this situation is that the changes need to appear on both the `develop` and `feature/otx` branches. In the `feature/otx` branch, the `external` directory has been deprecated and will be removed in favor of the `algorithms` directory. In this case, task related changes should be made to `algorithms/anomaly`. The tests should be located in the `tests` directory in the root folder, not in the task implementation directory. (I'm not sure where exactly to place the tests as there is no test design for this yet.)

Do you have any ideas regarding how to merge this PR in the simplest way, @ashwinvaidya17 and @AlexanderDokuchaev?

**Comment by AlexanderDokuchaev** [comment]
@samet-akcay @ashwinvaidya17 
I created https://github.com/openvinotoolkit/training_extensions/pull/1330 to develop branch.
In this PR i added changes to `training_extensions/otx/*` and `training_extensions/tests/*'.

File: external/anomaly/tasks/inference.py [hunk 6]
```diff
@@ -291,8 +291,13 @@ def save_model(self, output_model: ModelEntity) -> None:
...
         output_model.set_data("pixel_threshold", self.model.pixel_threshold.value.cpu().numpy().tobytes())
-        output_model.set_data("min", self.model.min_max.min.cpu().numpy().tobytes())
-        output_model.set_data("max", self.model.min_max.max.cpu().numpy().tobytes())
+        if hasattr(self.model, "min_max"):
+            output_model.set_data("min", self.model.min_max.min.cpu().numpy().tobytes())
+            output_model.set_data("max", self.model.min_max.max.cpu().numpy().tobytes())
+        else:
+            logger.warning(
+                "The model was not trained before saving. This will lead to incorrect normalization of the heatmaps."
+            )
```
Comment (samet-akcay) [inline external/anomaly/tasks/inference.py:300]: @AlexanderDokuchaev, can you make these changes in [otx/algorithms/anomaly/tasks/inference](https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/anomaly/tasks/inference.py#L306-L307)?

Once `feature/otx` is merged, `external` directory will be removed.

File: external/anomaly/tests/anomaly_common.py [hunk 7]
```diff
@@ -18,6 +18,10 @@
 from collections import namedtuple
 from copy import deepcopy
 from typing import List, Type
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from nncf.torch.nncf_network import NNCFNetwork
```
Comment (samet-akcay) [inline external/anomaly/tests/anomaly_common.py:24]: And will you be able to move all these tests to somewhere in`tests` directory in the root `~/training_extensions/tests`?

Comment (samet-akcay) [reply external/anomaly/tests/anomaly_common.py:24]: The idea is that we would like to combine all the tests to a single location, rather than scattering them around different places in the repo

File: external/anomaly/tasks/inference.py (hunk 6 above)
Comment (samet-akcay) [reply external/anomaly/tasks/inference.py:300]: Never mind, they are there. Thanks

File: external/anomaly/tests/anomaly_common.py (hunk 7 above)
Comment (samet-akcay) [reply external/anomaly/tests/anomaly_common.py:24]: Same as above, thanks for addressing this.

**Review by samet-akcay — APPROVED** [review]
lgtm, thanks!
